src/libraries/System.Private.CoreLib/src/System/IO/FileStatus.Unix.cs

```diff
@@ -321,7 +316,24 @@ private unsafe void SetAccessOrWriteTime(string path, DateTimeOffset time, bool
             }
 #endif
             Interop.CheckIo(Interop.Sys.UTimensat(path, buf), path, InitiallyDirectory);
-            _initializedFileCache = -1;
+
+            // On OSX platforms, when the modification time is less than the creation time, the creation time
+            // is broken; so these api calls revert the creation time when it shouldn't be set (since we're
+            // setting modification time and access time here). checkCreationTime is only true on OSX
+            // platforms. allowFallbackToLastWriteTime is ignored on non OSX platforms.
+            bool updateCreationTime = checkCreationTime && (_fileCache.Flags & Interop.Sys.FileStatusFlags.HasBirthTime) != 0 &&
+                                        (buf[1].TvSec < _fileCache.BirthTime || (buf[1].TvSec == _fileCache.BirthTime && buf[1].TvNsec < _fileCache.BirthTimeNsec));
```

Comment:
Nanoseconds is the value of how many nanoseconds we are past that second (see `UnixTimeSecondsToNanoseconds`), also we need it for the modification date specifically. It will still run very fast how it currently is anyway I think.